File: SS/Labs/Lab2/1_arithmetic_expression.c
// Program to recognize arithmetic expression and identify identifiers & operators
#include <stdio.h>
#include <string.h>
#include <ctype.h>

char identifiers[20][20];
char operators[20];
int id_count = 0, op_count = 0;

int main() {
    char expr[100];
    char token[20];
    int i, j, valid = 1;
    
    printf("Enter arithmetic expression: ");
    fgets(expr, sizeof(expr), stdin);
    expr[strlen(expr)-1] = '\0';  // remove newline
    
    for (i = 0; expr[i] != '\0'; i++) {
        // Skip spaces
        if (expr[i] == ' ' || expr[i] == '\n')
            continue;
        
        // Check for operators
        if (expr[i] == '+' || expr[i] == '-' || expr[i] == '*' || 
            expr[i] == '/' || expr[i] == '%') {
            operators[op_count] = expr[i];
            op_count++;
        }
        // Check for identifiers (starts with letter)
        else if (isalpha(expr[i])) {
            j = 0;
            while (isalnum(expr[i])) {
                token[j] = expr[i];
                j++;
                i++;
            }
            token[j] = '\0';
            i--;  // adjust for loop increment
            strcpy(identifiers[id_count], token);
            id_count++;
        }
        // Skip numbers
        else if (isdigit(expr[i])) {
            while (isdigit(expr[i]))
                i++;
            i--;
        }
        // Skip parentheses
        else if (expr[i] == '(' || expr[i] == ')') {
            continue;
        }
        else {
            valid = 0;
            break;
        }
    }
    
    if (valid) {
        printf("\nVALID arithmetic expression\n");
        
        printf("\nIdentifiers: ");
        for (i = 0; i < id_count; i++)
            printf("%s ", identifiers[i]);
        
        printf("\nOperators: ");
        for (i = 0; i < op_count; i++)
            printf("%c ", operators[i]);
        printf("\n");
    }
    else {
        printf("\nINVALID expression\n");
    }
    
    return 0;
}

File: SS/Labs/PythonLabs/Lab2/1_arithmetic_expression.py
# Lab 2 — Arithmetic Expression Analyzer
# Identifies identifiers and operators in a valid arithmetic expression

def analyze(expr):
    identifiers = []
    operators = []
    valid = True
    i = 0

    while i < len(expr):
        ch = expr[i]

        if ch in ' \t\n':
            i += 1
            continue

        if ch in '+-*/%':
            operators.append(ch)
            i += 1

        elif ch.isalpha():
            j = i
            while i < len(expr) and expr[i].isalnum():
                i += 1
            identifiers.append(expr[j:i])

        elif ch.isdigit():
            while i < len(expr) and expr[i].isdigit():
                i += 1

        elif ch in '()':
            i += 1

        else:
            valid = False
            break

    return valid, identifiers, operators

def main():
    expr = input("Enter arithmetic expression: ")
    valid, identifiers, operators = analyze(expr)

    if valid:
        print("\nVALID arithmetic expression")
        print("Identifiers:", " ".join(identifiers))
        print("Operators:  ", " ".join(operators))
    else:
        print("\nINVALID expression")

if __name__ == "__main__":
    main()
